utils/include/detray/simulation/landau_distribution.hpp

```diff
@@ -0,0 +1,293 @@
+/** Detray library, part of the ACTS project (R&D line)
+ *
+ * (c) 2023 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s).
+#include "detray/definitions/math.hpp"
+
+// System include(s).
+#include <limits>
+#include <random>
+
+namespace detray {
+
+/// Landau distribution based on CERNLIB G110
+/// Computer Physics Communications 31(1984) 97—111)
+template <typename scalar_t>
+class landau_distribution {
+
+    public:
+    using scalar_type = scalar_t;
+
+    /// Generate a random number following a Landau distribution
+    /// with location parameter mu and scale parameter sigma:
+    ///      Landau( (x-mu)/sigma )
+    /// Note that mu is not the mpv(most probable value) of the Landa
+    /// distribution and sigma is not the standard deviation of the distribution
+    /// which is not defined. For mu  =0 and sigma=1, the mpv = -0.22278
+    ///
+    /// The Landau random number generation is implemented using the
+    /// function landau_quantile(x,sigma), which provides
+    /// the inverse of the landau cumulative distribution.
+    /// landau_quantile has been converted from CERNLIB ranlan(G110).
+    ///
+    /// Reference[1]: ROOT TRandom.cxx
+    /// Reference[2]: ACTS LandauDistribution.cxx
+    template <typename generator_t>
+    scalar_type operator()(generator_t &generator, const scalar_type location,
+                           const scalar_type scale) const {
+        const auto z = std::uniform_real_distribution<scalar_type>()(generator);
+        // LANDAU quantile : algorithm from CERNLIB G110 ranlan
+        // Converted by Rene Brun from CERNLIB routine ranlan(G110),
+        // Moved and adapted to QuantFuncMathCore by B. List 29.4.2010
+        return location + scale * quantile(z);
+    }
+
+    private:
+    scalar_type quantile(const scalar_type z) const {
+
+        static const scalar_type f[982] = {
+            0.f,        0.f,        0.f,        0.f,        0.f,
+            -2.244733f, -2.204365f, -2.168163f, -2.135219f, -2.104898f,
```

Comment:
I am not sure encoding these values in single precision is what we want, since they seem to specify a large number of digits. Since this is in the simulation package, can we not run the calculations here in double precision and then cast at the right spots to ```scalar_type```? This is not used in core or any performance critical device code, right?

Reply:
OK